osu.Game/Rulesets/Mods/ModAdaptiveSpeed.cs

```diff
@@ -0,0 +1,183 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Audio;
+using osu.Framework.Audio.Track;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Audio;
+using osu.Framework.Utils;
+using osu.Game.Beatmaps;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public class ModAdaptiveSpeed : Mod, IApplicableToRate, IApplicableToDrawableHitObject, IApplicableToBeatmap, IApplicableToHUD
+    {
+        // use a wider range so there's still room for adjustment when the initial rate is extreme
+        private const double fastest_rate = 2.5f;
+        private const double slowest_rate = 0.4f;
+
+        /// <summary>
+        /// Adjust track rate using the average speed of the last x hits
+        /// </summary>
+        private const int average_count = 6;
+
+        public override string Name => "Adaptive Speed";
+
+        public override string Acronym => "AS";
+
+        public override string Description => "Let track speed adapt to you.";
+
+        public override ModType Type => ModType.Fun;
+
+        public override double ScoreMultiplier => 1;
+
+        public override Type[] IncompatibleMods => new[] { typeof(ModRateAdjust), typeof(ModTimeRamp) };
+
+        [SettingSource("Initial rate", "The starting speed of the track")]
+        public BindableNumber<double> InitialRate { get; } = new BindableDouble
+        {
+            MinValue = 0.5,
+            MaxValue = 2,
+            Default = 1,
+            Value = 1,
+            Precision = 0.01
+        };
+
+        [SettingSource("Adjust pitch", "Should pitch be adjusted with speed")]
+        public BindableBool AdjustPitch { get; } = new BindableBool
+        {
+            Default = true,
+            Value = true
+        };
+
+        public BindableNumber<double> SpeedChange { get; } = new BindableDouble
+        {
+            Default = 1,
+            Value = 1,
+            Precision = 0.01
+        };
+
+        private ITrack track;
+        private HUDOverlay overlay;
+
+        private readonly List<double> recentRates = Enumerable.Range(0, average_count).Select(_ => 1d).ToList();
```

Comment:
```suggestion
        private readonly List<double> recentRates = Enumerable.Repeat(1d, average_count).ToList();
```

Shame that C# doesn't have a baked-in deque structure but alas.